src/cascadia/TerminalSettingsModel/ActionMap.cpp

```diff
@@ -59,31 +59,93 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+        std::unordered_map<InternalActionID, Model::Command> InboxActions;
+        winrt::com_ptr<implementation::ActionMap> foundParent{ nullptr };
+        for (const auto parent : _parents)
+        {
+            for (const auto [_, cmd] : parent->_ActionMap)
+            {
+                if (cmd.Origin() != OriginTag::InBox)
+                {
+                    // only one parent contains all the inbox actions and that parent contains only inbox actions,
+                    // so if we found a non-inbox action we can just skip to the next parent
+                    break;
+                }
+                foundParent = parent;
+                break;
+            }
+        }
+
+        if (foundParent)
         {
-            // ActionMap should never point to nullptr
-            FAIL_FAST_IF_NULL(maskingPair->second);
+            for (const auto [_, cmd] : foundParent->_ActionMap)
+            {
+                InboxActions.emplace(Hash(cmd.ActionAndArgs()), cmd);
+            }
+        }
+
+        // now, look through our _ActionMap for commands that
+        // - had an ID generated for them
+        // - do not have a name/icon path
+        // - have a hash that matches a command in the inbox actions
+        std::unordered_set<winrt::hstring> IdsToRemove;
+        for (const auto [userID, userCmd] : _ActionMap)
+        {
+            const auto userCmdImpl{ get_self<Command>(userCmd) };
 
-            // masking actions cannot contain nested or invalid commands,
-            // so we can just return it directly.
-            return maskingPair->second;
+            // Note we don't need to explicitly check for the origin tag here since we only generate IDs for user actions,
+            // so if we ID was generated it means this is a user action
+            if (userCmdImpl->IdWasGenerated() && !userCmdImpl->HasName() && userCmd.IconPath().empty())
+            {
+                const auto userActionHash = Hash(userCmd.ActionAndArgs());
+                if (const auto inboxCmd = InboxActions.find(userActionHash); inboxCmd != InboxActions.end())
+                {
+                    for (auto [key, cmdID] : _KeyMap)
+                    {
+                        // for any of our keys that point to the user action, point them to the inbox action instead
+                        if (cmdID == userID)
+                        {
+                            _KeyMap.insert_or_assign(key, inboxCmd->second.ID());
+
+                            // register the keys with the inbox action
+                            inboxCmd->second.RegisterKey(key);
+                        }
+                    }
+
+                    // add this ID to our set of IDs to remove
+                    IdsToRemove.insert(userID);
+                }
+            }
         }
 
+        // now, remove the commands with the IDs we found
+        for (const auto id : IdsToRemove)
+        {
+            _ActionMap.erase(id);

```

Comment:
FYI you can replace this with a `std::erase_if()` where you simply return a bool whether to remove any given element.